capsules/src/usb/ctap.rs

```diff
@@ -0,0 +1,391 @@
+/// Plus it matches: https://chromium.googlesource.com/chromiumos/platform2/+/master/u2fd/u2fhid.cc
+static REPORT_DESCRIPTOR: &'static [u8] = &[
+    0x06, 0xD0, 0xF1, // HID_UsagePage ( FIDO_USAGE_PAGE ),
+    0x09, 0x01, // HID_Usage ( FIDO_USAGE_CTAPHID ),
+    0xA1, 0x01, // HID_Collection ( HID_Application ),
+    0x09, 0x20, // HID_Usage ( FIDO_USAGE_DATA_IN ),
+    0x15, 0x00, // HID_LogicalMin ( 0 ),
+    0x26, 0xFF, 0x00, // HID_LogicalMaxS ( 0xff ),
+    0x75, 0x08, // HID_ReportSize ( 8 ),
+    0x95, 0x40, // HID_ReportCount ( HID_INPUT_REPORT_BYTES ),
+    0x81, 0x02, // HID_Input ( HID_Data | HID_Absolute | HID_Variable ),
+    0x09, 0x21, // HID_Usage ( FIDO_USAGE_DATA_OUT ),
+    0x15, 0x00, // HID_LogicalMin ( 0 ),
+    0x26, 0xFF, 0x00, // HID_LogicalMaxS ( 0xff ),
+    0x75, 0x08, // HID_ReportSize ( 8 ),
+    0x95, 0x40, // HID_ReportCount ( HID_OUTPUT_REPORT_BYTES ),
+    0x91, 0x02, // HID_Output ( HID_Data | HID_Absolute | HID_Variable ),
+    0xC0, // HID_EndCollection
+];
+
+static REPORT: ReportDescriptor<'static> = ReportDescriptor {
+    desc: REPORT_DESCRIPTOR,
+};
+
+static SUB_HID_DESCRIPTOR: &'static [HIDSubordinateDescriptor] = &[HIDSubordinateDescriptor {
+    typ: DescriptorType::Report,
+    len: REPORT_DESCRIPTOR.len() as u16,
+}];
+
+static HID_DESCRIPTOR: HIDDescriptor<'static> = HIDDescriptor {
+    hid_class: 0x0110,
+    country_code: HIDCountryCode::NotSupported,
+    sub_descriptors: SUB_HID_DESCRIPTOR,
+};
+
+/// Implementation of the CTAP HID (Human Interface Device)
+pub struct CtapHid<'a, U: 'a> {
+    /// Helper USB client library for handling many USB operations.
+    client_ctrl: ClientCtrl<'a, 'static, U>,
+
+    /// 64 byte buffers for each endpoint.
+    buffers: [Buffer64; N_ENDPOINTS],
+
+    /// A holder reference for the TX buffer we are transmitting from.
+    tx_buffer: TakeCell<'static, [u8]>,
+    /// The number of bytes the client has asked us to send. We track this so we
+    /// can pass it back to the client when the transmission has finished.
+    tx_len: Cell<usize>,
+    /// Where in the `tx_buffer` we need to start sending from when we continue.
+    tx_offset: Cell<usize>,
+    /// The TX client to use when transmissions finish.
+    tx_client: OptionalCell<&'a dyn uart::TransmitClient>,
+
+    /// A holder for the buffer to receive bytes into. We use this as a flag as
+    /// well, if we have a buffer then we are actively doing a receive.
+    rx_buffer: TakeCell<'static, [u8]>,
+    /// How many bytes the client wants us to receive.
+    rx_len: Cell<usize>,
+    /// How many bytes we have received so far.
+    rx_offset: Cell<usize>,
+    /// The RX client to use when RX data is received.
+    rx_client: OptionalCell<&'a dyn uart::ReceiveClient>,
+}
+
+impl<'a, U: hil::usb::UsbController<'a>> CtapHid<'a, U> {
+    pub fn new(
+        controller: &'a U,
+        vendor_id: u16,
+        product_id: u16,
+        strings: &'static [&'static str; 3],
```

Comment:
These should be typed (e.g., pass a struct that has three strings as named fields).

Reply:
Then this would differ from the CDC device and the ClientCtrl::new() arguments.